README.md
<!--
    THIS FILE IS GENERATED FROM crates-io.md, badges.md AND LICENSE.md.
    DO NOT EDIT IT DIRECTLY.
-->

# float_eq 

[![Build](https://github.com/jtempest/float_eq-rs/actions/workflows/build.yml/badge.svg)](https://github.com/jtempest/float_eq-rs/actions/workflows/build.yml)
[![codecov](https://codecov.io/gh/jtempest/float_eq-rs/branch/main/graph/badge.svg?token=QR20064LIR)](https://codecov.io/gh/jtempest/float_eq-rs)
[![crate](https://img.shields.io/crates/v/float_eq.svg)](https://crates.io/crates/float_eq)
[![documentation](https://docs.rs/float_eq/badge.svg)](https://docs.rs/float_eq)

Compare IEEE floating point primitives, structs and collections for equality.

This crate provides an API with a focus on making the choices of comparison 
algorithm(s) and tolerances intuitive to implementers and maintainers, and of
providing clear output for debugging and development iteration.

This readme is a quick tour of the crate. For introductory material, guides and
discussion see [the float_eq guide].

## Usage

Add this to your cargo.toml:

```
[dependencies]
float_eq = "1"
```

And, if you're using the 2015 edition, this to your crate root:

```rust
extern crate float_eq;
```

Then, you can import items with `use`:

```rust
use float_eq::{assert_float_eq, float_eq};
```

## Comparisons

This crate provides boolean comparison operations:

```rust
if (float_eq!(y_pos, 0.0, abs <= 0.000_1)) {
    //...
}
```

And asserts:

```rust
const RECIP_REL_TOL: f32 = 0.000_366_210_94;
assert_float_eq!(x.recip(), 10.0, r2nd <= RECIP_REL_TOL);
```

Using absolute tolerance, relative tolerance or ULPs based [comparison
algorithms].

## Composite types

Composite types may implement the provided extension traits to be compared on a
field-by-field basis:

```rust
let a = Complex32 { re: 2.0, im: 4.000_002 };
let b = Complex32 { re: 2.000_000_5, im: 4.0 };

assert_float_eq!(a, b, ulps <= ComplexUlps32 { re: 2, im: 4 });
```

...and if they are homogeneous, with a uniformly applied tolerance across all
fields:

```rust
assert_float_eq!(a, b, ulps_all <= 4);
```

Arrays of any size are supported:

```rust
let a = [1.0, -2.0, 3.0];
let b = [-1.0, 2.0, 3.5];
assert_float_eq!(a, b, abs <= [2.0, 4.0, 0.5]);
assert_float_eq!(a, b, abs_all <= 4.0);
```

As are tuples up to size 12 (inclusive):

```rust
let a = (1.0f32, 2.0f64);
let b = (1.5f32, -2.0f64);
assert_float_eq!(a, b, r2nd <= (0.5, 2.0));
```

Many standard and core types like `Vec` are supported:

```rust
let a = vec![1.0, -2.0, 3.0];
let b = vec![-1.0, 2.0, 3.5];
assert_float_eq!(a, b, rmax <= vec![2.0, 2.0, 0.25]);
assert_float_eq!(a, b, rmax_all <= 2.0);
```

There are blanket trait impls for comparing mutable and immutable reference
types, the contents of `Cell`, `RefCell`, `Rc`, `Arc` and `Box` instances, as
well as for slices, `Option`, `Vec`, `VecDeque`, `LinkedList`, `BTreeMap` and
`HashMap`.

## Derivable

The extension traits may be derived for non-generic structs and tuple structs:

```rust
#[derive_float_eq(
    ulps_tol = "PointUlps",
    ulps_tol_derive = "Clone, Copy, Debug, PartialEq",
    debug_ulps_diff = "PointUlpsDebugUlpsDiff",
    debug_ulps_diff_derive = "Clone, Copy, Debug, PartialEq",
    all_tol = "f64"
)]
#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

let a = Point { x: 1.0, y: -2.0 };
let c = Point { 
    x: 1.000_000_000_000_000_9, 
    y: -2.000_000_000_000_001_3
};
assert_float_eq!(a, c, ulps <= PointUlps { x: 4, y: 3 });
assert_float_eq!(a, c, ulps_all <= 4);
```

## Error messages

Asserts provide additional useful context information. For example:

```rust
assert_float_eq!(4.0f32, 4.000_008, rmax <= 0.000_001);
```

Panics with this error message:

```
thread 'main' panicked at 'assertion failed: `float_eq!(left, right, rmax <= t)`
        left: `4.0`,
       right: `4.000008`,
    abs_diff: `0.000008106232`,
   ulps_diff: `Some(17)`,
    [rmax] t: `0.000004000008`', assert_failure.rs:15:5
```

Where `[rmax] t` shows the tolerance value that the absolute difference was
compared against after being appropriately scaled.

## Optional features

This crate can be used without the standard library (`#![no_std]`) by disabling
the default `std` feature. Use this in `Cargo.toml`:

```
[dependencies.float_eq]
version = "1"
default-features = false
```

Other optional features:
- **derive** — provides custom derive macros for all traits.
- **num** — blanket trait impls for `num::Complex` where it is instanced with a
  compatible type.

## Related efforts

The [`approx`], [`float-cmp`], [`assert_float_eq`] and [`is_close`] crates provide
similar floating point comparison capabilities to `float_eq`. The [`almost`] crate
divides its API into comparison of floats against zero and non-zero values. The
[`efloat`] crate provides an `f32` equivalent type that tracks the maximum
possible error bounds that may have occured due to rounding.

The [`ieee754`] crate is not a comparison library, but provides useful
functionality for decomposing floats into their component parts, iterating over
representable values and working with ULPs directly, amoung other things.

## Contributing 

Constructive feedback, suggestions and contributions welcomed, please
[open an issue].

## Changelog

Release information is available in [CHANGELOG.md](CHANGELOG.md).

[comparison algorithms]: https://jtempest.github.io/float_eq-rs/book/background/float_comparison_algorithms.html
[open an issue]: https://github.com/jtempest/float_eq-rs/issues/
[the float_eq guide]: https://jtempest.github.io/float_eq-rs/book/introduction.html
[`almost`]: https://crates.io/crates/almost
[`approx`]: https://crates.io/crates/approx
[`assert_float_eq`]: https://crates.io/crates/assert_float_eq
[`efloat`]: https://crates.io/crates/efloat
[`float-cmp`]: https://crates.io/crates/float-cmp
[`ieee754`]: https://crates.io/crates/ieee754
[`is_close`]: https://docs.rs/is_close/latest/is_close/

<br>

#### License

<sup>
Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
</sup>

<br>

<sub>
Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in float_eq by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.
</sub>

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2022-10-12

### Fixed
- Fixed a bug where field visibility was not respected when deriving the Ulps
  type.
  - https://github.com/jtempest/float_eq-rs/issues/24
- Fixed documentation to refer to `ComplexUlps32` rather than the incorrect
  `Complex32Ulps`.
  - https://github.com/jtempest/float_eq-rs/issues/23

## [1.0.0] - 2022-06-02
Added some resource links to the documentation and promoted the crate to
release version 1.0.0, as it has been stable for some time now.

## [0.7.0] - 2021-10-03
Bumped up the version number since this release includes breaking API changes.

### Changed
- Derived types no longer have a default set of derived traits, use the newly
  added `ulps_tol_derive` and `debug_ulps_diff_derive` options to `#[float_eq]`
  to provide a list of traits to be derived instead.
- The "derive" feature no longer requires the "full" feature from `syn`.
- Separated out and clarified the documentation for implementing the various
  extension traits.

## [0.6.1] - 2021-07-31

### Changed
- Clearer documentation for the `FloatEqDebugUlpsDiff` and `FloatEqUlpsTol`
  traits.
- Updated derive tests to match current compiler output.
- Switched to using github actions and codecov.io for continuous integration
  and coverage.

## [0.6.0] - 2021-05-03
Bumped up the version number since this release includes breaking API changes.

### Changed
- Renamed many ambiguous uses of epsilon/eps/max_diff to tolerance/tol, which is
  the mathematical term of art.
- Implementation of traits for arrays of any size using const generics.
- Updated `num-complex` dependency to 0.4.
- Reworked the documentation, added a tutorial.

### Fixed
- Safer behaviour when initialising debug output for array trait impls,
  including the removal of a source of undefined behaviour.

## [0.5.0] - 2020-08-24
Bumped up the version number since this release includes breaking API changes
to the extension traits.

### Changed
- Renamed `FloatEqDebug` to `AssertFloatEq`, and similarly `FloatEqAllDebug` to 
  `AssertFloatEqAll`. This is to provide a more obvious link between these
  traits and the assert capability. The `FloatDiff` trait has been folded into
  `AssertFloatEq` both for clarity and because it is geared towards debug
  context and was not as general as the name implied.
- Expanded `rel` checks into `rmax` (aka `rel`), `rmin`, `r1st` and `r2nd` to
  select which operand to scale epsilon's granularity to.
- Significant rewrite of the documentation to clarify aspects of comparison and
  to have a slightly more streamlined structure with better defined sections.
- Updated `num-complex` dependency to 0.3.

## [0.4.1] - 2020-06-22

### Added
- The `derive_float_eq` helper macro to simplify deriving traits.

### Fixed
- Update README usage documentation to the correct version, correctly documented
  the "num" feature.

## [0.4.0] - 2020-06-19
Bumped up the version number since this release includes breaking API changes.

### Added
- `#[derive]` support for all traits on tuple structs and structs with named
  fields.
- The `FloatUlps` trait, which is a way to syntactically tie a floating point
  type to its ULPs representation. The prefered way to express this is now to
  implement `FloatUlps for T` and then use `Ulps<T>` for the ULPs type. The
  other traits have been reworked to reflect this.
- Blanket trait impls for comparing slices, Option, Vec, VecDeque, LinkedList,
  BTreeMap and HashMap.

### Changed
- Blanket array impls now allow for comparison of arrays of different types.
- `FloatEqAll::Epsilon` is now `AllEpsilon` and `FloatEqAllDebug::DebugEpsilon`
  is now `AllDebugEpsilon` to help avoid ambiguity.

## [0.3.1] 2020-06-03

### Added
- Blanket trait impls for comparing mutable and immutable reference types and
  the contents of Cell, RefCell, Rc, Arc and Box instances.
- More descriptive documentation for combining types of check.

## [0.3.0] - 2020-05-30
Bumped up the version number since this release includes breaking API changes.

### Added
- The `FloatEqAll` and `FloatEqAllDebug` traits, which allow for `abs_all`,
  `rel_all` and `ulps_all` to be used in the `float_eq` and `assert_float_eq`
  families of macros, which take a single epsilon value to be applied uniformly
  across all fields being compared. The behaviour of `FloatEq` has now changed
  to apply epsilon values structurally, on a per-field basis, since that is a
  more general behaviour (e.g. all tuples may sensibly be `FloatEq` but not
  `FloatEqAll`). If existing checks against arrays or `num::Complex` break, 
  switching to use the `_all` variants ought to fix them.
- Support for tuples of up to size 12 (inclusive).
- Documentation on how to interpret assert error messages.

### Changed
- ULPs checks now treat `NaN` values as not equal, to match the default partial
  equality behaviour of floats.
- Equality of infinities is now consistent, both internally and with respect to
  general IEEE floating point behaviour.
- `FloatEq` now more specifically means equality based on a structurally defined
  epsilon type. See the notes on `FloatEqAll` above.
- `FloatDiff` is much more rigorously defined, in particular `ulps_diff` now
  returns `Option<T>`, see the API documentation for details.
- `FloatEq`'s `DiffEpsilon` is now `Epsilon`, and `UlpsDiffEpsilon` is now
  `UlpsTol`. This reduces visual noise around the usage and harmonizes the
  naming with `FloatEqDebug`'s associated types.
- Directed docs.rs to build documentation for all features.

## [0.2.0] - 2020-04-12
Bumped up the version number since this release includes breaking API changes.

### Added
- Implementation of traits for arrays of size 0 to 32 (inclusive) where the type

src/arrays.rs
use crate::{FloatDiff, FloatEq, FloatEqAll, FloatEqAllDebug, FloatEqDebug};
use core::mem::MaybeUninit;

// arrays
//TODO: Should this be publically available for users to conditionally implement
// support if they need it?
macro_rules! impl_float_eq_traits_for_array {
    ($n:literal) => {
        impl<T: FloatDiff> FloatDiff for [T; $n] {
            type AbsDiff = [<T as $crate::FloatDiff>::AbsDiff; $n];
            type UlpsDiff = [<T as $crate::FloatDiff>::UlpsDiff; $n];

            #[inline]
            fn abs_diff(&self, other: &Self) -> Self::AbsDiff {
                let mut result: Self::AbsDiff = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].abs_diff(&other[i])
                }
                result
            }

            #[inline]
            fn ulps_diff(&self, other: &Self) -> Self::UlpsDiff {
                let mut result: Self::UlpsDiff = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].ulps_diff(&other[i])
                }
                result
            }
        }

        impl<T: FloatEq> FloatEq for [T; $n] {
            type Epsilon = [T::Epsilon; $n];
            type UlpsEpsilon = [T::UlpsEpsilon; $n];

            #[inline]
            fn eq_abs(&self, other: &Self, max_diff: &Self::Epsilon) -> bool {
                for i in 0..$n {
                    if !self[i].eq_abs(&other[i], &max_diff[i]) {
                        return false;
                    }
                }
                true
            }

            #[inline]
            fn eq_rel(&self, other: &Self, max_diff: &Self::Epsilon) -> bool {
                for i in 0..$n {
                    if !self[i].eq_rel(&other[i], &max_diff[i]) {
                        return false;
                    }
                }
                true
            }

            #[inline]
            fn eq_ulps(&self, other: &Self, max_diff: &Self::UlpsEpsilon) -> bool {
                for i in 0..$n {
                    if !self[i].eq_ulps(&other[i], &max_diff[i]) {
                        return false;
                    }
                }
                true
            }
        }

        impl<T: FloatEqAll> FloatEqAll for [T; $n] {
            type Epsilon = T::Epsilon;
            type UlpsEpsilon = T::UlpsEpsilon;

            #[inline]
            fn eq_abs_all(&self, other: &Self, max_diff: &Self::Epsilon) -> bool {
                self.iter()
                    .zip(other.iter())
                    .all(|(a, b)| a.eq_abs_all(b, max_diff))
            }

            #[inline]
            fn eq_rel_all(&self, other: &Self, max_diff: &Self::Epsilon) -> bool {
                self.iter()
                    .zip(other.iter())
                    .all(|(a, b)| a.eq_rel_all(b, max_diff))
            }

            #[inline]
            fn eq_ulps_all(&self, other: &Self, max_diff: &Self::UlpsEpsilon) -> bool {
                self.iter()
                    .zip(other.iter())
                    .all(|(a, b)| a.eq_ulps_all(b, max_diff))
            }
        }

        impl<T: FloatEqDebug> FloatEqDebug for [T; $n] {
            type DebugEpsilon = [T::DebugEpsilon; $n];
            type DebugUlpsEpsilon = [T::DebugUlpsEpsilon; $n];

            #[inline]
            fn debug_abs_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::Epsilon,
            ) -> Self::DebugEpsilon {
                let mut result: Self::DebugEpsilon = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_abs_epsilon(&other[i], &max_diff[i])
                }
                result
            }

            #[inline]
            fn debug_rel_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::Epsilon,
            ) -> Self::DebugEpsilon {
                let mut result: Self::DebugEpsilon = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_rel_epsilon(&other[i], &max_diff[i])
                }
                result
            }

            #[inline]
            fn debug_ulps_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::UlpsEpsilon,
            ) -> Self::DebugUlpsEpsilon {
                let mut result: Self::DebugUlpsEpsilon =
                    unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_ulps_epsilon(&other[i], &max_diff[i])
                }
                result
            }
        }

        impl<T: FloatEqAllDebug> FloatEqAllDebug for [T; $n] {
            type DebugEpsilon = [T::DebugEpsilon; $n];
            type DebugUlpsEpsilon = [T::DebugUlpsEpsilon; $n];

            #[inline]
            fn debug_abs_all_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::Epsilon,
            ) -> Self::DebugEpsilon {
                let mut result: Self::DebugEpsilon = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_abs_all_epsilon(&other[i], max_diff)
                }
                result
            }

            #[inline]
            fn debug_rel_all_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::Epsilon,
            ) -> Self::DebugEpsilon {
                let mut result: Self::DebugEpsilon = unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_rel_all_epsilon(&other[i], max_diff)
                }
                result
            }

            #[inline]
            fn debug_ulps_all_epsilon(
                &self,
                other: &Self,
                max_diff: &Self::UlpsEpsilon,
            ) -> Self::DebugUlpsEpsilon {
                let mut result: Self::DebugUlpsEpsilon =
                    unsafe { MaybeUninit::uninit().assume_init() };
                for i in 0..$n {
                    result[i] = self[i].debug_ulps_all_epsilon(&other[i], max_diff)
                }
                result
            }
        }
    };
}

impl_float_eq_traits_for_array!(0);
impl_float_eq_traits_for_array!(1);
impl_float_eq_traits_for_array!(2);
impl_float_eq_traits_for_array!(3);
impl_float_eq_traits_for_array!(4);
impl_float_eq_traits_for_array!(5);
impl_float_eq_traits_for_array!(6);
impl_float_eq_traits_for_array!(7);
impl_float_eq_traits_for_array!(8);
impl_float_eq_traits_for_array!(9);
impl_float_eq_traits_for_array!(10);
impl_float_eq_traits_for_array!(11);
impl_float_eq_traits_for_array!(12);
impl_float_eq_traits_for_array!(13);
impl_float_eq_traits_for_array!(14);
impl_float_eq_traits_for_array!(15);
impl_float_eq_traits_for_array!(16);
impl_float_eq_traits_for_array!(17);
impl_float_eq_traits_for_array!(18);
impl_float_eq_traits_for_array!(19);
impl_float_eq_traits_for_array!(20);
impl_float_eq_traits_for_array!(21);
impl_float_eq_traits_for_array!(22);
impl_float_eq_traits_for_array!(23);
impl_float_eq_traits_for_array!(24);
impl_float_eq_traits_for_array!(25);
impl_float_eq_traits_for_array!(26);
impl_float_eq_traits_for_array!(27);
impl_float_eq_traits_for_array!(28);
impl_float_eq_traits_for_array!(29);
impl_float_eq_traits_for_array!(30);
impl_float_eq_traits_for_array!(31);
impl_float_eq_traits_for_array!(32);
